[RFC] Add systemd configuration support to cgconfig.conf

kamalesh-babulal opened the pull request:
This patch series introduces systemd configurations to `cgconfig.conf`
and support to parse them. This will help the users to create `systemd slice/scope`
when system boots, and also allow them to set the resource allocations to the created
`slice/scope`.  A sample configuration would look like this:
```
systemd {
        slice = testing;
        scope = testing;
}

systemd {
        slice = testing;
        scope = testing1;
        setdefault = yes;
}

group testing.slice {
        cpu {
               cpu.weight = 10;
        }
}
```
and it also adds support to the tools to consider the default slice as the cgroup root
for all operations performed by the tool, instead of the controller mount point.  This
will help users on working on the default `slice/scope` hierarchy they created as the
new cgroup root .  The user can also ignore the default hierarchy setting by using the
new switch `-b` added to all of the tools.

Werkov commented:
The examples are slightly confusing (everything named `testing`, I'd suggest more distinctive names in the example for inner/leaf nodes), it's not clear what the `group testing.slice` refers to from the two sections above.

Also note, that [systemd delegation](https://systemd.io/CGROUP_DELEGATION/) is defined only for service and scopes, not slices (internal nodes). I understand your approach is similar (not identical) to *i-like-continents* scenario in that document.

kamalesh-babulal commented:
> The examples are slightly confusing (everything named testing, I'd suggest more distinctive names in the example for inner/leaf nodes), it's not clear what the group testing.slice refers to from the two sections above. Also note, that [systemd delegation](https://systemd.io/CGROUP_DELEGATION/) is defined only for service and scopes, not slices (internal nodes). I understand your approach is similar (not identical) to i-like-continents scenario in that document.

@Werkov Thank you for reviewing the patch.  If the example `cgconfig.conf` is used, the `systemd-cgls` output will look like:
```
[...]
│   └─ 728 /usr/bin/abrt-dump-journal-oops -fxtD
└─testing.slice 
  ├─testing.scope …
  │ └─ 986 libcgroup_systemd_idle_thread
  └─testing1.scope …
    └─ 987 libcgroup_systemd_idle_thread
```
I will try and come up with a better names for them and add more comments with references to the sample/example to make it more self-explanatory.

drakenclimber commented on include/libcgroup/config.h:
```diff
@@ -29,13 +29,57 @@ extern "C" {
+ * Parse the systemd delegation settings from the configuration file
+ * and get the default delegation cgroup name, if set.
+ *
+ * @param pathname Name of the configuration file to load.
```
How does this function behave if the default isn't set?

kamalesh-babulal replied:
@drakenclimber Thank you so much for reviewing the patches. It will return the path without appending the default slice cgroup name.

drakenclimber commented on src/api.c:
```diff
@@ -1752,6 +1774,33 @@ char *cg_build_path_locked(const char *name, char *path, const char *type)
+				 * avoid using systemd delegate slice as path
+				 * prefix.
+				 */
+				if (name[0] == '/') {
```
This feels kinda klunky.  So we build a path above, but if the first char is `/`, we'll override that path and put a new one in instead?

drakenclimber commented on src/api.c:
```diff
@@ -1752,6 +1774,33 @@ char *cg_build_path_locked(const char *name, char *path, const char *type)
+				 * prefix.
+				 */
+				if (name[0] == '/') {
+					if (cg_namespace_table[i]) {
```
Also, somewhat of a corner case, but tools like cgget used to allow the user to get settings in the root cgroup via `/` as the cgroup name.  How will that work here?

drakenclimber commented on tests/Makefile.am:
```diff
@@ -1,4 +1,5 @@
 if WITH_TESTS
-SUBDIRS = $(DIST_SUBDIRS)
+SUBDIRS = gunit
+#SUBDIRS = $(DIST_SUBDIRS)
```
This feels like an oops :)

kamalesh-babulal replied:
My bad, was testing some bug with Gunit.

drakenclimber commented on src/tools/cgget.c:
```diff
@@ -746,6 +759,14 @@ int main(int argc, char *argv[])
 		goto err;
 
+	if (mode & MODE_SYSTEMD_DELEGATE) {
+		ret = cgroup_get_default_slice_name(CGCONFIG_CONF_FILE);
```
What if the user is utilizing the cgconfigconfdir, `/etc/cgconfig.d`?

kamalesh-babulal replied:
@drakenclimber Thanks for all the corner cases and functionality, I had missed. I will re-work them and share the changes sooner.

kamalesh-babulal commented on src/config.c:
```diff
@@ -608,6 +616,96 @@ void cgroup_config_cleanup_namespace_table(void)
 	memset(&config_namespace_table, 0, sizeof(struct cg_mount_table_s) * CG_CONTROLLER_MAX);
 }
 
+/**
```
Agreed, will remove them.

kamalesh-babulal commented on src/config.c:
```diff
@@ -887,6 +985,49 @@ static int config_validate_namespaces(void)
+	if (!_cgroup_systemd_default_slice_name)
+		return 1;
+
+	scope_opts.delegated = 1;
```
:) Yes, I should be calling it instead.

drakenclimber commented on src/config.c:
```diff
@@ -608,6 +616,96 @@ void cgroup_config_cleanup_namespace_table(void)
+
+	if (strncmp(config, "slice\0", 6)   != 0 &&
+	    strncmp(config, "scope\0", 6)   != 0 &&
+	    strncmp(config, "setdefault\0", 11) != 0)
```
Thinking out loud here:

What if we did something like this?  Then the user could pick any arbitrary relative path.  Thoughts?
```
systemd {
        slice = testing;
        scope = testing;
}

default_cgroup_path {
    testing.slice/testing.scope
}
```

drakenclimber replied:
And I don't really love the `default_cgroup_path` name, so I'm open to ideas there.

kamalesh-babulal commented on include/libcgroup/config.h:
```diff
@@ -29,13 +29,57 @@ extern "C" {
+/*
+ * cgroup systemd settings.
+ */
+struct cgroup_systemd_opts {
```
Given that, `systemd` will be heavily used and it will be enabled by default in the configurable option, we can very well remove it, on the other hand, we would wait for the feedback from the community and remove it from the release + 1 having configurable support for systemd, after fixing any bugs reported.

kamalesh-babulal commented on src/config.c:
```diff
@@ -1061,6 +1204,39 @@ static int cgroup_parse_config(const char *pathname)
+ *
+ * @param pathname Name of the configuration file to load.
+ */
+int cgroup_get_default_slice_name(const char * const pathname)
```
Agreed, should it be renamed to `cgroup_set_default_slice_name` or another other better name?

drakenclimber replied:
That sounds good to me

kamalesh-babulal commented:
**V2 Changes**
- renamed `cgroup_get_default_slice_name()` to `cgroup_set_default_systemd_cgroup()`
- Improved the naming of the systemd configuration examples
- The default delegated cgroup path is `<slice name>.slice/<scope name>.scope`, whereas in the previous version it was just   `<slice name>.slice/`
- Introduced `WITH_SYSTEMD CFLAGS` for `src/*` and `src/tools/*` that would guard for systemd code  across the sources
- Introduced `/var/run/libcgroug/systemd` to store the default delegated slice/scope name, this addresses the issue of multiple` cgconfig.conf` files passed via `-d` option to the `cgconfigparser`.
- Introduced `systemd_default_cgroup_lock` to avoid race conditions on file operation on `systemd_default_cgroup_file` (`/var/run/libcgroup/systemd)`
- `cg_build_path_locked()` - redid the logic cgroup name starting with a '/'  where the '/<cgroup name> and '/' are honored while constructing the path using the systemd_default_cgroup.
- `cg_build_path_locked()` - fix the `snprintf` warning using a temporary variable to construct the path.
- `tools/*` will no longer pass the CGCONFIG_CONF_FILE, instead read from the  `systemd_default_cgroup_file` (`/var/run/libcgroup/systemd`)
- `tools/*` will only display options of overriding the `systemd_default_cgroup`,  as well as constructing a path with `systemd_default_cgroup_file` iff configured using systemd.
- Other improvements in error checkings and added more comments.

drakenclimber reviewed (commented):
I think this is looking pretty good.  Seems like we're going to need a ton of tests to verify this doesn't break all the various cgget/cgset options.  There are already a lot of permutations there, and this adds even more :)

kamalesh-babulal commented on include/libcgroup/config.h:
```diff
@@ -29,13 +29,56 @@ extern "C" {
  * and applications can benefit from them.
  */
 
+/*
```
@drakenclimber Thanks a lot for the review. Sure, make sense to move them.

kamalesh-babulal commented on src/config.c:
```diff
@@ -887,6 +908,7 @@ static int config_validate_namespaces(void)
 	return error;
 }
 
+
```
Will fix it in the next version.

kamalesh-babulal commented on src/api.c:
```diff
@@ -1684,8 +1684,36 @@ char *cg_build_path_locked(const char *name, char *path, const char *type)
+	 * effectively overriding the systemd_default_cgroup.  If the name
+	 * is '/', path is cgroup root or systemd_default_cgroup root.
+	 */
+	if (strlen(systemd_default_cgroup) && name && name[0] == '/' && name[1] != '\0')
```
There are chances the user doesn't set the `systemd->setdefault` in the configuration, hence `systemd_default_cgroup` can be zero length.  override it only if its set, else `tmp_systemd_default_cgroup` will be zero-length string.

drakenclimber replied:
Sounds like a really good thing to add a test around :)

kamalesh-babulal commented on src/api.c:
```diff
@@ -1684,8 +1684,36 @@ char *cg_build_path_locked(const char *name, char *path, const char *type)
+	if (strlen(systemd_default_cgroup) && name && name[0] == '/' && name[1] != '\0')
+		tmp_systemd_default_cgroup[0] = '\0';
+	else
+		snprintf(tmp_systemd_default_cgroup, len, "%s/", (systemd_default_cgroup));
```
The extra byte is required for the `/`, to keep GCC happy `snprintf()` space allocation. The global `systemd_default_group` has only one `/` (slice/scope)

kamalesh-babulal commented on src/config.c:
```diff
@@ -2044,6 +2044,87 @@ static int config_create_slice_scope(char * const tmp_systemd_default_cgroup)
+		goto err;
+	}
+
+	snprintf(tmp, sizeof(tmp), "%s/cpu.stat", path);
```
Agree, using `cgroup.procs` is reliable over `cpu.stat`.

kamalesh-babulal commented on src/config.c:
```diff
@@ -2044,6 +2044,87 @@ static int config_create_slice_scope(char * const tmp_systemd_default_cgroup)
+	fclose(systemd_def_cgrp_f);
+	/* at the minimum it should be <n>.slice/<n>.scope == 15 */
+	if (len < 15) {
+		cgroup_dbg("Malformed system default cgroup %s", systemd_default_cgroup);
```
Will fix it in the next iteration.

kamalesh-babulal commented on src/tools/cgget.c:
```diff
@@ -583,6 +599,19 @@ static int fill_empty_controller(struct cgroup * const cg, struct cgroup_control
+		path[FILENAME_MAX - 1] = '\0';
+	}
+	path_len = strlen(path);
+	if (cg->name[0] == '/' && strncmp(path + (path_len - 6), ".slice/", 6) == 0) {
```
The first comparison is made for the `7` bytes, because the path ends with a `/` and that needs to be taken into account  but after trimming the scope cgroup from the path using `dirname()`, which also deletes the `/`, hence the second comparison is done for 6 characters but you are right, that the comparison is misleading and `.slice` should be used instead.

kamalesh-babulal commented on src/tools/cgget.c:
```diff
@@ -583,6 +599,19 @@ static int fill_empty_controller(struct cgroup * const cg, struct cgroup_control
 
 	path_len = strlen(path);
+#ifdef WITH_SYSTEMD
+	if (cg->name[0] == '/' && strncmp(path + (path_len - 7), ".scope/", 7) == 0) {
```
Sorry, will add comments here.

kamalesh-babulal commented:
**V3 Changes**
Addressed the review  comments from Tom  and a few notable changes are:
- Moved all the definitions to `systemd.h`.
- Use `cgroup.procs` instead of  `cpu.stat` to check if the cgroup exists.
- Added documentation at a couple of places.
- Introduce `cgget` changes to `cgxget` too.

drakenclimber commented:
I think this is looking pretty good.  Let's get some tests around it and see what they find.

A while back I prototyped a couple commits that help with testing cgget in this patchset.  They may be a good start:
https://github.com/drakenclimber/libcgroup/commit/7bdb5f14926bea396ec0b344a548e274c06218a4
https://github.com/drakenclimber/libcgroup/commit/2afe401bcf324a6c07fda269b03dd601297c6476

It looks like I didn't fully flesh out the `ignore_systemd` flag in the first patch, but it's close.

kamalesh-babulal commented on src/api.c:
```diff
@@ -1674,7 +1679,51 @@ static char *cg_concat_path(const char *pref, const char *suf, char *path)
+		cgroup_err("Failed to allocate memory for tmp_systemd_default_cgroup.\n");
+		goto out;
+	}
+	memset(tmp_systemd_default_cgroup, 0 , sizeof(*tmp_systemd_default_cgroup));
```
@Werkov Thanks for the review. Sure, will replace it using `calloc()`

kamalesh-babulal commented on src/config.c:
```diff
@@ -1807,3 +1855,287 @@ int cgroup_config_create_template_group(struct cgroup *cgroup, char *template_na
+{
+	struct cgroup_systemd_opts *curr = cgroup_systemd_opts_tail;
+
+	if (strncmp(config, "slice\0", 6) == 0)
```
I wanted to use `strncmp()` to limit the check to `slice`, given that it's read from the configuration file, the user can always give a wrong/lengthier configuration option.  On the other hand, agree that it's unlikely, will change these constructions with `strcmp()`.

kamalesh-babulal commented:
**V4 Changes**:
- Addressed the review comments from Michal Koutný
- Added systemd config option `pid` that lets the users add an existing task as the scope task in place of the `idle thread`
- Added test cases from Tom Hromatka
- Added test case to test `pid` based on the test cases written by Tom Hromatka

drakenclimber commented on src/tools/cgset.c:
```diff
@@ -139,8 +144,16 @@ int main(int argc, char *argv[])
+			ignore_default_systemd_delegate_slice = 1;
+			break;
+#else
+	while ((c = getopt_long (argc, argv, "r:hb", long_options, NULL)) != -1) {
```
Typo - should be `r:h`

kamalesh-babulal replied:
:( sorry for typos will fix them

drakenclimber commented on src/tools/cgdelete.c:
```diff
@@ -150,9 +154,16 @@ int main(int argc, char *argv[])
+			ignore_default_systemd_delegate_slice = 1;
+			break;
+#else
+	while ((c = getopt_long(argc, argv, "rhg:b", long_options, NULL)) > 0) {
```
typo - should be `rhg:`

drakenclimber commented on src/tools/cgexec.c:
```diff
@@ -63,9 +67,16 @@ int main(int argc, char *argv[])
 	memset(cgroup_list, 0, sizeof(cgroup_list));
-
+#ifdef WITH_SYSTEMD
 	while ((c = getopt_long(argc, argv, "+g:sh", longopts, NULL)) > 0) {
```
I think this should be `+g:shb`, correct?

drakenclimber commented on tests/ftests/054-sudo-g_flag_controller_only_systemd-v1.py:
```diff
@@ -0,0 +1,158 @@
+    # Let's remove the existing default systemd cgroup file
+    # so that the temp cgroup we create is at the know path.
+    if os.path.exists(DEFAULT_SYSTEMD_CGROUP_NAME):
+        os.remove(DEFAULT_SYSTEMD_CGROUP_NAME)
```
This feels like belt and suspenders code to me - which isn't a bad thing.

But, it feels weird in this test.  Is a user expected to do this?  Shouldn't calling `configparser()` automatically cause this to be overwritten.  And if `configparser()` should overwrite this, are we potentially hiding a bug because we aren't doing things the same way a user would?

kamalesh-babulal replied:
Agreed with all the points, let me remove them so that when fails we  know  something  went wrong, which is the purpose of the test :)

drakenclimber commented on tests/ftests/054-sudo-g_flag_controller_only_systemd-v1.py:
```diff
@@ -0,0 +1,158 @@
+    # The job of the tmp cgroup is to create a task, that will be
+    # used by the systemd as pid for the scope.
+    Cgroup.create(config, CONTROLLER, TMP_CGNAME)
+    config.process.create_process_in_cgroup(config, CONTROLLER, TMP_CGNAME)
```
Why are you creating a process in a cgroup only to have it be moved to the newly-created scope?  Wouldn't calling `config.create_process()` be simpler?

kamalesh-babulal replied:
yeah, it will be way more elegant.

drakenclimber commented on src/config.c:
```diff
@@ -91,6 +92,26 @@ static struct cgroup_string_list *template_files;
+ * cgroup_systemd_opts is dynamically allocated, hence having head
+ * and tail helps in traversing.
+ */
+static struct cgroup_systemd_opts *cgroup_systemd_opts_head;
```
Unless I'm missing something, I think we need to initialize these to NULL.

kamalesh-babulal replied:
@drakenclimber Thank you so much for the review.
Isn't `static` storage initialized to a null pointer unless initialized explicitly?
```
$ echo -e "struct { int num;}a;\n struct a *foo;\n int main() {\n if (foo) { return 1;}\n return 0;\n}" | gcc -x c - -o -
$ echo $?
0
```

drakenclimber commented on tests/ftests/systemd.py:
```diff
@@ -27,3 +29,91 @@ def is_delegated(config, scope_name):
+                ctrl_mnt = Cgroup.get_controller_mount_point(controller)
+                if 'Unknown version' not in ctrl_mnt:
+                    break
+            else:
```
nifty use of the `for/else` loop :)

drakenclimber commented on tests/ftests/systemd.py:
```diff
@@ -27,3 +29,91 @@ def is_delegated(config, scope_name):
+            controllers = ['cpu', 'memory', 'io', 'pids', 'cpuset']
+            for controller in controllers:
+                ctrl_mnt = Cgroup.get_controller_mount_point(controller)
+                if 'Unknown version' not in ctrl_mnt:
```
I don't think this `if` is buying you anything.  Unknown version and unknown mount point are raised as Exceptions, and thus I don't think this `if` is protecting you.

drakenclimber commented on tests/ftests/systemd.py:
```diff
@@ -27,3 +29,91 @@ def is_delegated(config, scope_name):
+            if ignore_systemd:
+                raise IndexError('Unsupported controller \'None\'')
+
+            controllers = ['cpu', 'memory', 'io', 'pids', 'cpuset']
```
Does it make sense to have a list of controllers here, or does it make more sense to make `get_controller_mount_point()` smarter and be able to handle a `None` controller?

kamalesh-babulal replied:
@drakenclimber smart :) Let me implement it

kamalesh-babulal replied:
Something like below:
```
--- a/tests/ftests/cgroup.py
+++ b/tests/ftests/cgroup.py
@@ -741,6 +741,10 @@ class Cgroup(object):
             for line in mntf.readlines():
                 mnt_path = line.split()[1]
 
+                if ctrl_name is None:
+                    if line.split()[0] == 'cgroup2':
+                        return mnt_path
+
                 if line.split()[0] == 'cgroup':
                     for option in line.split()[3].split(','):
                         if option == ctrl_name:
@@ -758,6 +762,9 @@ class Cgroup(object):
                 mnt_path = line.split()[1]
 
                 if line.split()[0] == 'cgroup2':
+                    if ctrl_name is None:
+                        return mnt_path
+
                     ctrl_file = os.path.join(line.split()[1],
                                              'cgroup.controllers')
 
@@ -774,7 +781,16 @@ class Cgroup(object):
 
     @staticmethod
     def get_controller_mount_point(ctrl_name):
-        vers = CgroupVersion.get_version(ctrl_name)
+        if ctrl_name is None:
+            mode = int(Cgroup.get_cgroup_mode(None))
+            # map return modes:
+            # CGROUP_MODE_LEGACY/CGROUP_MODE_HYBRID ->  CgroupVersion.CGROUP_V1
+            # CGROUP_MODE_UNIFIED -> CgroupVersion.CGROUP_V2
+            if mode >= 2:
+                mode = mode - 1
+            vers = CgroupVersion(mode)
+        else:
+            vers = CgroupVersion.get_version(ctrl_name)
 
         if vers == CgroupVersion.CGROUP_V1:
             return Cgroup.__get_controller_mount_point_v1(ctrl_name)
```

drakenclimber commented on tests/ftests/060-sudo-cgconfigparser-systemd.py:
```diff
@@ -0,0 +1,194 @@
+                            'Unexpected error {}, while parsing configuration:'
+                            '\n{}'.format(re.stdout, configuration[0])
+                        )
+                cause = '\n'.join(filter(None, [cause, tmp_cause]))
```
This is a cool use of `filter()`.  Nice :)

drakenclimber commented on tests/ftests/065-sudo-systemd_cgclassify-v1.py:
```diff
@@ -0,0 +1,196 @@
+        cause = 'Changing group of pid {} erroneously succeeded'.format(OTHER_PIDS)
+
+    # classify the task from the non-systemd scope cgroup to systemd scope
+    # cgroup (should fail due to wrong path).
```
nitpick - I think this comment isn't right.

kamalesh-babulal replied:
Does the following re-phrasing make sense?
```      
# This should fail because of trying to migrate a task to SYSTEMD_CGNAME
# under the wrong path from OTHER_CGNAME.
```

drakenclimber replied:
It's a copy-paste of the previous comment.  Would this be correct?
```
# classify the task from the systemd scope cgroup to the non-systemd scope
# cgroup (should fail due to wrong path).
```

kamalesh-babulal replied:
Agreed, re-reading it, it's not obvious. Does the more verbose version make sense?
```
    # classify a task from the non-systemd scope cgroup (OTHER_CGNAME) to systemd                   
    # scope cgroup (SYSTEMD_CGNAME).  Migration should fail due to the incorrect                    
    # destination cgroup path that gets constructed, without the systemd slice/scope                
    # when ignore_systemd is set) 
```

drakenclimber commented on tests/ftests/066-sudo-systemd_cgclassify-v2.py:
```diff
@@ -0,0 +1,213 @@
+        cause = 'Changing group of pid {} erroneously succeeded'.format(OTHER_PIDS)
+
+    # classify the task from the non-systemd scope cgroup to systemd scope
+    # cgroup (should fail due to wrong cgroup path).
```
nitpick - same comment as on the previous test.  I think this comment is a copy/paste error :)

drakenclimber commented on tests/ftests/067-sudo-systemd_cgexec-v1.py:
```diff
@@ -0,0 +1,199 @@
+    return result, cause
+
+
+def create_process_get_pid(config, CGNAME, SLICE='', ignore_systemd=False):
```
nitpick - you probably don't want to override the global variable named `SLICE`.  It'll work as you hope, but it's kinda a maintenance nightmare

kamalesh-babulal replied:
Agree, will rename it `SLICENAME` to avoid confusion.

drakenclimber commented on tests/ftests/067-sudo-systemd_cgexec-v1.py:
```diff
@@ -0,0 +1,199 @@
+    if result == consts.TEST_FAILED:
+        return result, cause
+
+    # classify the task from the non-systemd scope cgroup to systemd scope
```
I don't understand the need to classify the processes to another cgroup.  Isn't the point of this test to run cgexec?  If so, wouldn't a clearer validation simply be to verify that the process(es) were created in the requested cgroups?

kamalesh-babulal replied:
Agreed, just checking for process creation under the right cgroup should be sufficient and the migration test is not required here

drakenclimber commented on tests/ftests/068-sudo-systemd_cgexec-v2.py:
```diff
@@ -0,0 +1,216 @@
+    return result, cause
+
+
+def create_process_get_pid(config, CGNAME, SLICE='', ignore_systemd=False):
```
same nitpick comment as previous test

drakenclimber commented on tests/ftests/069-sudo-systemd_cgxget-cpu-settings-v1.py:
```diff
@@ -0,0 +1,169 @@
+    for i in cgrps:
+        for entry in TABLE:
+            Cgroup.xset(config, cgname=i, setting=entry[0], value=entry[1],
+                        version=entry[2], ignore_systemd=cgrps[i])
```
Dictionaries are only ordered on Python 3.7 and above.  Do we care that this code won't work on older versions of Python?  Should we enforce that?

drakenclimber replied:
I'm specifically referring to the `cgrps[i]` snippet of code

kamalesh-babulal replied:
The unordered Dictionaries will not harm there, because `cgrps[SYSTEMD_CGNAME] = > ignore_systemd=False` and `cgrps[OTHER_CGNAME] => ignore_systemd=True` always no matter the order, and `xset` will find the right path of the cgroup.

drakenclimber replied:
Thanks.  Makes sense

drakenclimber commented on src/config.c:
```diff
@@ -1807,3 +1857,236 @@ int cgroup_config_create_template_group(struct cgroup *cgroup, char *template_na
+	 */
+	if (strcmp(config, "slice") != 0 &&
+	    strcmp(config, "scope") != 0 &&
+	    strcmp(config, "setdefault") != 0 &&
```
My apologies up front.... :)

Now that I think about it more, would naming this parameter `default` (rather than `setdefault`) make more sense?  We didn't name the other parameters `setslice`, `setscope`, etc. but they're obviously setting something.

kamalesh-babulal replied:
I completely agree, in fact, I had thought about renaming it to `default` will rename it.

kamalesh-babulal replied:
actually, `default` can't be used as its a keyword, does `delegate` sound better?

drakenclimber replied:
Argh... thanks for digging.  Let's stick with `setdefault` then.  `delegate` is likely an option we will need to add later, but it should apply to the `opts->delegate` field within the systemd options structure.

kamalesh-babulal replied:
Ah, Make sense, will leave it unchanged.

drakenclimber commented on src/tools/lscgroup.c:
```diff
@@ -47,6 +47,9 @@ static void usage(int status, const char *program_name)
 	info("  -g <controllers>:<path>	Control group to be ");
```
nitpick - First line of the commit comment mentions `cgcreate` :)

kamalesh-babulal replied:
copy, and paste typos, I tried being cautious by still, sorry :(

drakenclimber commented on src/tools/cgclassify.c:
```diff
@@ -41,6 +41,9 @@ static void usage(int status, const char *program_name)
 	info("  --cancel-sticky		cgred daemon change pidlist and children tasks\n");
```
nitpick - `cgcreate` is mentioned twice in the commit comment :)

drakenclimber commented on src/tools/cgexec.c:
```diff
@@ -49,11 +49,15 @@ static void usage(int status, const char *program_name)
 	info("  -h, --help			Display this help\n");
```
nitpick - first line of the commit comment mentions `cgcreate` :)

drakenclimber commented on tests/ftests/systemd.py:
```diff
@@ -27,3 +29,91 @@ def is_delegated(config, scope_name):
+    # systemd path and checks if the cgroup exists. The function can build
+    # path for all three cgroup setup modes, including empty cgroup v2.
+    @staticmethod
+    def cgroup_exists(config, ctrl_name, cgroup_name, ignore_systemd=False):
```
Hmmm, thinking out loud here...  Currently you aren't using `config` but would we need it if this function were run within a container?  I think so

kamalesh-babulal replied:
Sure, that will make it more `bash` than `Python` :)

drakenclimber replied:
I don't have a strong preference here.  If you haven't changed it, leave it as is.  This is internal to the tests and we can change its parameters later with no repercussions.

kamalesh-babulal replied:
Sure, I have re-written it :)

drakenclimber commented on tests/ftests/systemd.py:
```diff
@@ -27,3 +29,91 @@ def is_delegated(config, scope_name):
+    # cgroup and cgroup_exists(), which builds the cgroup path and returns
+    # True if the cgroup exists, False otherwise
+    @staticmethod
+    def create_and_validate(config, ctrl_name, cgroup_name, ignore_systemd=False):
```
`create()` is located in the Cgroup class.  Would it make sense to have `create_and_validate()` next to it in the same class?  (It seems like we should consider converting some of the `create()` calls in other tests to `create_and_validate()` and thus putting this in Cgroup class might make more sense.

drakenclimber commented on tests/ftests/systemd.py:
```diff
@@ -27,3 +29,91 @@ def is_delegated(config, scope_name):
+    # systemd path and checks if the cgroup exists. The function can build
+    # path for all three cgroup setup modes, including empty cgroup v2.
+    @staticmethod
+    def cgroup_exists(config, ctrl_name, cgroup_name, ignore_systemd=False):
```
Also, would it make more sense to have this function in the Cgroup class, and it could just be called `Cgroup.exists()`?

kamalesh-babulal replied:
Something like this with `get_controller_mount_point()` support for `None`
```
+    @staticmethod
+    def exists(config, ctrl_name, cgroup_name, ignore_systemd=False):
+        ctrl_mnt = Cgroup.get_controller_mount_point(ctrl_name)
+
+        if (ignore_systemd):
+            cgrp_path = os.path.join(ctrl_mnt, cgroup_name)
+        else:
+            try:
+                cmd = [ 'cat', '/run/libcgroup/systemd']
+                if config.args.container:
+                    slice_scope_path = config.container.run(cmd, shell_bool=True)
+                else:
+                    slice_scope_path = Run.run(cmd, shell_bool=True)
+            except RunError:
+                    return False
+
+            cgrp_path = os.path.join(ctrl_mnt, slice_scope_path, cgroup_name)
+
+        proc_file =  os.path.join(cgrp_path, 'cgroup.procs')
+        cmd = ['cat', proc_file]
+        if config.args.container:
+            try:
+                exists = config.container.run(cmd, shell_bool=True)
+            except RunError:
+                return False
+        else:
+            try:
+                exists = Run.run(cmd, shell_bool=True)
+            except RunError:
+                return False
+
+        return True
```

kamalesh-babulal commented:
### V6 Changes:
- Addressed the review comments
- Included the newly added test cases to the `tests/ftests/Makefile.am`

drakenclimber commented:
Pushed to main.  Thanks so much @kamalesh-babulal for all the hard work and time!

I'll merge it into release-3.0 in a bit.

drakenclimber commented:
Pushed to release-3.0.  Let me know if you notice anything awry.  Thanks, all!
